File: Merge-sort/java/MergeSort.java
public class MergeSort {
    public static void mergeSort(int[] arr, int l , int m, int r){
        int s1 = m - l + 1;
        int s2 = r - m;
        int aL[] = new int[s1];
        int aR[] = new int[s2];
        int i = 0;
        int j = 0;
        int k = l;

        for (i = 0; i < s1; ++i)
            aL[i] = arr[l + i];
        for (j = 0; j < s2; ++j)
            aR[j] = arr[m + 1 + j];

        i = 0;
        j = 0;

        while (i < s1 && j < s2) {
            if (aL[i] <= aR[j]) {
                arr[k] = aL[i];
                i++;
            }
            else {
                arr[k] = aR[j];
                j++;
            }
            k++;
        }
        while (i < s1) {
            arr[k] = aL[i];
            i++;
            k++;
        }
        while (j < s2) {
            arr[k] = aR[j];
            j++;
            k++;
        }
    }
    public static void sort(int arr[], int l, int r)
    {
        if (l < r) {
            int m = l + (r - l) / 2;
            sort(arr, l, m);
            sort(arr, m + 1, r);
            mergeSort(arr, l, m, r);
        }
    }

    public static void main (String[] args){
        int[] arr = {64, 34, 25, 12, 22, 11, 90};
        System.out.println("Unsorted array:");
        for (int i = 0; i <arr.length - 1; i++)
            System.out.print(arr[i] + " ");
        System.out.println();
        sort(arr, 0, arr.length - 1);
        System.out.println("Sorted array:");
        for (int i = 0; i <arr.length - 1; i++)
            System.out.print(arr[i] + " ");
        System.out.println();
    }
}

File: Merge-sort/python/MergeSort.py
def mergeSort(arr, l, m, r):
    s1 = m - l + 1
    s2 = r - m
    aL = [0] * s1
    aR = [0] * s2
    i = 0
    j = 0
    k = l

    for i in range(s1):
        aL[i] = arr[l + i]
    for j in range(s2):
        aR[j] = arr[m + 1 + j]

    i = 0
    j = 0

    while i < s1 and j < s2:
        if aL[i] <= aR[j]:
            arr[k] = aL[i]
            i += 1
        else:
            arr[k] = aR[j]
            j += 1
        k += 1

    while i < s1:
        arr[k] = aL[i]
        i += 1
        k += 1

    while j < s2:
        arr[k] = aR[j]
        j += 1
        k += 1
        
def sort(arr, l, r):
    if (l < r):
        m = l + (r - l) // 2
        sort(arr, l, m)
        sort(arr, m + 1, r)
        mergeSort(arr, l, m, r)

arr = [64, 34, 25, 12, 22, 11, 90]
print("Unsorted array:")
for i in range(len(arr)):
    print(arr[i], end=" ")
print()
sort(arr, 0, len(arr)-1)
print("Sorted array:")
for i in range(len(arr)):
    print(arr[i], end=" ")
print()